src/vcs/git/tag.rs

```diff
@@ -0,0 +1,160 @@
+// This file is part of radicle-surf
+// <https://github.com/radicle-dev/radicle-surf>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::vcs::git::{self, error::Error, reference::Ref, Author};
+use git2::Oid;
+use std::{convert::TryFrom, fmt, str};
+
+/// A newtype wrapper over `String` to separate out the fact that a caller wants
+/// to fetch a tag.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TagName(String);
+
+impl fmt::Display for TagName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl TryFrom<&[u8]> for TagName {
+    type Error = str::Utf8Error;
+
+    fn try_from(name: &[u8]) -> Result<Self, Self::Error> {
+        let name = str::from_utf8(name)?;
+        let short_name = match git::ext::try_extract_refname(name) {
+            Ok(stripped) => stripped,
+            Err(original) => original,
+        };
+        Ok(Self(String::from(short_name)))
+    }
+}
+
+impl From<TagName> for Ref {
+    fn from(other: TagName) -> Self {
+        Self::Tag { name: other }
+    }
+}
+
+impl TagName {
+    /// Create a new `TagName`.
+    pub fn new(name: &str) -> Self {
+        TagName(name.into())
+    }
+
+    /// Access the string value of the `TagName`.
+    pub fn name(&self) -> &str {
+        &self.0
+    }
+}
+
+/// The static information of a [`git2::Tag`].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Tag {
+    /// A light-weight git tag.
+    Light {
+        /// The Object ID for the `Tag`, i.e the SHA1 digest.
+        id: Oid,
+        /// The name that references this `Tag`.
+        name: TagName,
+    },
+    /// An annotated git tag.
+    Annotated {
+        /// The Object ID for the `Tag`, i.e the SHA1 digest.
+        id: Oid,
+        /// The name that references this `Tag`.
+        name: TagName,
+        /// The named author of this `Tag`, if the `Tag` was annotated.
```

Comment:
How can the two fields be optional given that this is the `Annotated` variant?

Reply:
Apparently, they can still be `None`: https://docs.rs/git2/0.13.6/git2/struct.Tag.html#method.tagger + https://docs.rs/git2/0.13.6/git2/struct.Tag.html#method.message